coordinator/src/node.rs

```diff
@@ -11,38 +13,127 @@ use dlc_manager::payout_curve::PayoutPoint;
 use dlc_manager::payout_curve::PolynomialPayoutCurvePiece;
 use dlc_manager::payout_curve::RoundingInterval;
 use dlc_manager::payout_curve::RoundingIntervals;
+use lightning::ln::channelmanager::ChannelDetails;
+use rust_decimal::prelude::ToPrimitive;
+use rust_decimal::Decimal;
 use std::sync::Arc;
 use std::time::SystemTime;
 use std::time::UNIX_EPOCH;
+use trade::cfd::calculate_long_liquidation_price;
 use trade::cfd::calculate_margin;
+use trade::cfd::calculate_short_liquidation_price;
+use trade::Direction;
 use trade::TradeParams;
 
+// todo: in case of leverage 1.0 the liquidation will be set to 21000000. I guess this will result
+// in issues when the overall upper boundary is smaller than that. It looks like the payout function
+// will never finish in that case. Setting this value temporarily to 2.0 so that the liquidation
+// price is below that total upper boundary.
+const COORDINATOR_LEVERAGE: f64 = 2.0;
+
 pub struct Node {
     pub inner: Arc<ln_dlc_node::node::Node>,
 }
 
 impl Node {
-    pub async fn trade(&self, trade_params: TradeParams) -> Result<()> {
-        // The coordinator always trades at a leverage of 1
-        let coordinator_leverage = 1.0;
+    pub async fn trade(&self, trade_params: &TradeParams) -> Result<()> {
+        // todo: eventually we will have to cover the following scenarios here.
+        // 1. open a completely new position (no pre-existing position)
+        // 2. extending a position, meaning that the position will be updated with a bigger
+        // quantity.
+        // 3. reducing a position, meaning that the position will be updated with
+        // a smaller quantity, and the trader will get a payout.
+        // 4. closing a position, if a the trade reversed the existing position.
 
-        let margin_coordinator = calculate_margin(
-            trade_params.execution_price,
-            trade_params.quantity,
-            coordinator_leverage,
-        );
-        let margin_trader = calculate_margin(
+        if self.is_position_closing(trade_params)? {
+            self.close_position(trade_params).await
+        } else {
+            self.open_position(trade_params).await
+        }
+    }
+
+    /// Checks if the existing position needs to be closed.
+    fn is_position_closing(&self, trade_params: &TradeParams) -> Result<bool> {
+        // FIXME: this needs to be replaced with a query from the database, matching the existing
+        // position on quantity and direction. If a "reversed" order was found we propose to
+        // collaboratively settle the dlc channel. If not the position either needs to be extended
+        // or reduced. e.g. if we have an existing long position of a quantity of 100 and we get a
+        // short position of a quantity of 75, we will simply reduce the existing dlc_channel to 75.
```

Comment:
```suggestion
        // short position of a quantity of 75, we will simply reduce the existing dlc_channel to 25.
```

I think `100 - 75 = 25` 😄
Note that other parameters of the position will also represent the aggregate of the orders, e.g. average leverage, average entry price and the resulting collateral (margin).